### ISSUE TITLE
designs as strategies

### ISSUE THREAD

**Issue by macartan** [DeclareDesign/DeclareDesign#229]
JC and I were chatting about designs in which the steps you take depend upon results from previous steps. For example gather more data, or nor, conditional on results of last analysis.

One could even imagine that such a process would have loops where some sequences are iterated until some goal is met (this came up in the context of thinking about declaring the Lieberman mixed methods strategy which has a loop of this form; though for designs-as-trees  issues see also  https://arxiv.org/abs/1307.1077 ; we might also confront when we declare adaptive designs). 

Can we imagine a way to do this? I imagine it might require a kind of design declaration that is separate from the concatenation of steps and more like a decision tree. 

Here is a kind of a version done "by hand," in which one seeks a population mean and resamples if the se is too high. One weakness of this, besides being very tailored, is that we cannot use diagnose_design on the resulting "design" --- the design here does produce a dataframe of estimates etc, but is not itself a design. Wonder if there would still be a way to have simulate_design and diagnose_design accept it.
  
```
n <- 5
P <- declare_population(N=1000, Y = rnorm(N))
S <- declare_sampling(n = n)
A <- declare_estimator(
  handler = tidy_estimator(function(data) {
    with(data, data.frame(se = sd(Y)/n^.5, n = length(Y)))}
    ))


looped_design <- function() {
  n <- 5
  design <- P+S+A 
  a <- get_estimates(design)

  while(a$se > .5) {
    n <- n+5
    design <- redesign(design, n = n)
    a <- get_estimates(design)
    }
  
  a
  }
```
This won't work:
```
simulate_design(looped_design)
```

But generating a simulations_df is not hard:

```reps <- replicate(10, expr = looped_design(), simplify =  FALSE)
simulations_df <- do.call("rbind", reps)
```

[An aside on this design; the design uses get_estimates but then in the next stage redraws the data rather than using the same data; ideally the same data would be passed on but the approach I imagine is pretty bulky, df <- draw_data(design), a <- design$A(df), P2 <- declare_population(data = df))...]

**Comment by macartan** [comment]
Thinking out loud: one might imagine something like:

D1  <- declare_decision(test, S2+A2, S3+A3)
design <- U + Y + Q  + Z  + R + D1

where D1 creates a kind of forking of a design that allows multiple tines (eg test here takes two possible values, but could allow more)

**Comment by nfultz** [comment]
For sequential designs, I think it's easier / cleaner to generate the whole thing up front, and use `subset` options on estimator/estimands instead of declare_sampling - once rows are dropped they aren't coming back.

Implementing declare_decision seems straightforward enough, but getting it compatible with all the options around executing designs (eg only run steps of certain types, different numbers of simulations at different steps, etc) would be hard.

**Comment by macartan** [comment]
Agreed on declare_sampling ; for such things declare_assignment and
conditional analysis safer

Interested if others have thoughts on implementation of declare_decision;
trees get complex very quickly

The other question is this: if this is not by hand, as in the example
above, can we find a way to still use the diagnose_design machinery on an
arbitrary function that generates a block of estimates (but which is not
itself a design)

**Comment by graemeblair** [comment]
I don't have any immediate ideas on how to do this better than the way @nfultz suggested, but open to exploring it.

The other kind of thing our framework has trouble with is the need to input features of the population during estimation, such as weights calculated from the population -- even though the dataset itself is the sampled data. An example of why you might want that is here: http://sekhon.berkeley.edu/papers/SATEtoPATT.pdf

Can you clarify your second question? Couldn't understand what a block of estimates is.

**Comment by macartan** [comment]
@graemeblair 
Think in such cases we need to keep the population or generate needed vars before subsetting. I think for a while we have felt it is hard to justify a distinction between assignment declaration and sampling declaration.

By the estimates dataframe I meant one can create a function that is not itself a design but that generates  results analogous to the results of a run of a design. It would be nice if such a function could be accepted by diagnose_design.

**Comment by graemeblair** [comment]
In the case of that paper at least, you need to have the sampled data and the population data at the same time to create the weights - they can't be created before sampling.

Ah got it, splitting that issue https://github.com/DeclareDesign/DeclareDesign/issues/230

**Comment by graemeblair** [comment]
One more thing related to this: I implemented in a long-ago version a way to do meta-analysis directly, i.e. 

`declare_meta_analysis(design_1, design_2, my_meta_estimator)`

There were two options, (1) meta analysis of the data so basically it runs each design and rbind's the data together and you do a meta-regression; and (2) meta analysis of the estimates so retains estimates and rbinds them.

**Comment by acoppock** [comment]
Re implementation, I think adding an explicit if statement via `declare_decision` is nice. 

Re using diagnose design on things that aren't design objects.... I *think* that diagnose design can take an estimates dataframe, no?

**Comment by graemeblair** [comment]
it can take a simulations df, but not a function that runs a design
